File: 백준/Gold/1407. 2로 몇 번 나누어질까/2로 몇 번 나누어질까.py

a, b = map(int, input().split())

def cnt(num):
    if num == 0:
        return 0
    
    elif num == 1:
        return 1

    elif num % 2 == 0: # 짝수일 때 1의 개수는 num//2개 + f(num // 2) * 2
        return cnt(num//2) * 2 + num // 2

    elif num % 2 == 1: #홀수일 때 f(num - 1) + 1개
        return cnt(num - 1) + 1

print(cnt(b) - cnt(a - 1))

File: 백준/Gold/1407. 2로 몇 번 나누어질까/2로 몇 번 나누어질까.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long a = Long.parseLong(st.nextToken());
        long b = Long.parseLong(st.nextToken());
        System.out.println(cnt(b) - cnt(a - 1));
    }

    static long cnt(long num){
        if (num == 1) return 1;
        if (num != 0 && num % 2 == 0) return cnt(num/2) * 2 + num/2;
        if (num % 2 == 1) return cnt(num - 1) + 1;
        return 0;
    }
}